### PR TITLE
Bug fixes

### LINKED ISSUE #110: Bug fixes
## Summary
The OpenAI Responses provider (`OpenAiResponsesAgent`, backing the gpt-5.5 / gpt-5.5-mini Copilot providers) constructed its own `IMessage`s and never called `.WithIds(options)`. Unlike `OpenAgent` / `AnthropicAgent`, it stamped only `GenerationId`, leaving `RunId` / `ParentRunId` / `ThreadId` null on every emitted message. In LmStreaming.Sample logs ~23% of outbound messages had `runId: null` (an entire gpt-5.5 conversation could be unstamped).

## Impact
`IMessage.RunId` is `[JsonIgnore(WhenWritingNull)]`, so null is omitted from client JSON. The client merge key `(kind, runId, generationId, messageOrderIdx)` loses a segment; fork/branch lineage (`ParentRunId`) and any run-keyed logic degrade. Currently masked only by per-turn `GenerationId` uniqueness.

## Fix
- Wrap the Responses agent's emitted stream so every message is stamped via `.WithIds(options)` (parity with the other providers); also stamp the non-streaming leftover `TextMessage`.
- `GenerationId` / BUG-H1 behavior unchanged — the stamp is idempotent when the run advertises a `GenerationId`.

## Files
- `src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs`
- `tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs` (regression coverage)

## Testing
OpenAiResponsesProvider.Tests 67/67 · LmMultiTurn.Tests 277/277 · CopilotLive.Tests 9/9.

## Out of scope (follow-ups)
- Secondary leak: straggler null `ToolCallResultMessage` on stamping providers.
- Pre-existing `IDE0370` build errors in unrelated test projects.

### PR DESCRIPTION
**Pull request by gautam-achieveai**
## Summary
Generic bug fixes bundled into one PR.

### 1. OpenAI Responses provider drops run identity (`runId == null`)
`OpenAiResponsesAgent` (gpt-5.5 / gpt-5.5-mini via Copilot) constructed its own `IMessage`s and never called `.WithIds(options)`, so every emitted message carried a null `RunId` / `ParentRunId` / `ThreadId` — ~23% of LmStreaming.Sample outbound logs showed `runId: null`. Now stamps the run ids on every emitted message, matching `OpenAgent` / `AnthropicAgent`. `GenerationId` / BUG-H1 behavior unchanged (the stamp is idempotent).

### 2. Streaming stops when switching conversations or refreshing
Switching conversations or refreshing tore down the WebSocket and, on return, only reloaded persisted REST history (everything marked completed) — the client never re-subscribed to the still-running backend run (the pooled agent keeps running after the client disconnects), and `SubscribeAsync` had no replay, so the visible stream froze. Fix:
- **Backend:** `MultiTurnAgentBase` keeps a bounded per-run replay buffer; `SubscribeAsync` atomically registers the subscriber and replays the in-flight run's already-published messages, then streams live ones (each message delivered exactly once — replay XOR live). New `GET /api/conversations/{threadId}/run-state` exposes the in-memory active-run signal.
- **Client:** `useChat.resumeStreamIfActive(threadId)` re-opens the WebSocket subscribe-only (no send) when the backend reports an in-flight run; `ChatLayout` calls it after loading history on switch / initial-load / refresh.

## Changes
- `OpenAiResponsesAgent`: `StampRunIds` wrapper applying `.WithIds(options)` to every emitted message.
- `MultiTurnAgentBase`: per-run replay buffer + atomic register/replay in `SubscribeAsync` (lock-guarded, no await under lock); buffer freed on run completion.
- `ConversationsController` + `ConversationDtos`: `GET {threadId}/run-state`.
- Client: `useChat.resumeStreamIfActive` + `conversationsApi.getRunState`; `ChatLayout` wiring; `buildStreamCallbacks`/`openStreamConnection` refactor (no behavior change to send).

## Testing
- OpenAiResponsesProvider.Tests **67/67** · LmMultiTurn.Tests **282/282** · client vitest **258/258** · vue-tsc clean.
- New coverage: `OpenAiResponsesAgentRunIdTests`; `MultiTurnAgentReplayTests` (replay contract, concurrent-subscribe exactly-once, multi-subscriber, buffer cap); `useChatResume.test.ts` (reconnect-and-resume, no-reconnect-when-idle, SSE skip, failed-open `isLoading` reset, stale-thread abort).
- Both fixes independently reviewed in fresh context; review findings addressed.

## Related Issues
Fixes #110

### CODE REVIEW

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Deep review of PR !111 — Bug fixes (RunId stamping for Responses provider)

## Context
Linked to #110. `OpenAiResponsesAgent` (GPT-5.5/GPT-5.5-mini via `CopilotResponsesAgentFactory`) constructed its own `IMessage`s without calling `.WithIds(options)`, leaving `RunId`/`ParentRunId`/`ThreadId` null on every emitted message (~23% of `LmStreaming.Sample` outbound logs had `runId==null`). The client merge key `(kind, runId, generationId, messageOrderIdx)` loses a segment when RunId is null — a latent identity bug masked only by per-turn `generationId` uniqueness.

## What Changed
Two fix sites:
1. **Streaming path** — wraps `EventStreamToMessages` output in a `StampRunIds` async helper applying `.WithIds(options)` per message (parity with `OpenAgent`/`AnthropicAgent`)
2. **Non-streaming leftover path** — applies `.WithIds(options)` on the synthesized `TextMessage` in `GenerateReplyAsync`

5 regression tests added via `ScriptedClient` covering: mixed stream, tool-call fallback, reasoning via output_item.done, full id propagation, and null-options.

## Verdict
✅ **Approve** — Fix is correct, well-scoped, consistent with existing agent patterns, and backed by solid tests. Two suggestions below.

## Review Agents Used
- `test-coverage-review` — found untested non-streaming path
- `performance-review` — no issues (StampRunIds overhead negligible vs network I/O)
- `class-design-simplifier` — StampRunIds wrapper justified (8 yield points), found test duplication

## Findings Summary
| # | Severity | Finding |
|---|----------|---------|
| 1 | **MEDIUM** | Missing test for `GenerateReplyAsync` leftover `TextMessage` path |
| 2 | LOW | Test 4 is near-duplicate of Test 1 |
| 3 | INFO | `UsageMessage.ParentRunId` null contract not positively asserted |
| 4 | INFO | `StampRunIds` wrapper vs inline — defensible design, no action |

See inline comments for findings 1 and 2.

File: tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            new ResponseOutputItemEvent
+            {
+                Type = ResponseEventTypes.OutputItemDone,
+                OutputIndex = 0,
+                Item = El("{\"type\":\"reasoning\",\"summary\":[{\"text\":\"deep thought\"}]}"),
+            },
+        ];
+
+    [Fact]
+    public async Task Every_emitted_message_carries_the_runs_run_id()
```
Comment (gautam-achieveai) [inline tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs:155]: **[Gautam's review bot]** [MEDIUM] **Missing test for `GenerateReplyAsync` leftover TextMessage path**

All 5 tests call `GenerateReplyStreamingAsync`, which exercises the `StampRunIds` wrapper (fix site #2 at line 132 of `OpenAiResponsesAgent.cs`). But `GenerateReplyAsync` has a **separate fix site** at line 100: when the stream emits `TextUpdateMessage` deltas without a finalized `TextMessage`, it synthesizes a new `TextMessage` with `.WithIds(options)` applied directly.

This path has zero test coverage. Someone could remove `.WithIds(options)` at line 100 (e.g., thinking `StampRunIds` already handles it) and no test would catch the regression. The leftover path fires on truncated/aborted streams — exactly the edge case where null-RunId identity bugs are hardest to diagnose.

Suggested test:
```csharp
private static ResponseEvent[] TextDeltasOnlyStream() =>
[
    new ResponseLifecycleEvent
    {
        Type = ResponseEventTypes.ResponseCreated,
        Response = El("{\"id\":\"resp_X\"}"),
    },
    new ResponseOutputTextDeltaEvent
    {
        Type = ResponseEventTypes.OutputTextDelta,
        OutputIndex = 0,
        Delta = "Hello",
    },
];

[Fact]
public async Task GenerateReplyAsync_leftover_text_carries_run_ids()
{
    const string runId = "run-123";
    using var agent = new OpenAiResponsesAgent("test",
        new ScriptedClient(TextDeltasOnlyStream()));
    var result = await agent.GenerateReplyAsync(
        [new TextMessage { Role = Role.User, Text = "go" }],
        new GenerateReplyOptions { RunId = runId, ParentRunId = "p", ThreadId = "t" });
    var text = result.OfType<TextMessage>().Single();
    text.RunId.Should().Be(runId);
    text.ParentRunId.Should().Be("p");
    text.ThreadId.Should().Be("t");
}
```

File: tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        }
+
+        var reasoning = messages.OfType<ReasoningMessage>().Single();
+        reasoning.RunId.Should().Be(runId);
+        reasoning.ParentRunId.Should().Be(parentRunId);
+        reasoning.ThreadId.Should().Be(threadId);
+    }
+
+    [Fact]
+    public async Task Tool_call_carries_full_run_thread_and_parent_ids()
```
Comment (gautam-achieveai) [inline tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs]: **[Gautam's review bot]** [LOW] **Test 4 is a near-duplicate of Test 1**

This test replays `MixedStream()` with identical options and asserts `RunId`/`ParentRunId`/`ThreadId` on the `ToolsCallMessage`. Test 1 already asserts these across **all** messages (including `ToolsCallMessage`) via `OnlyContain`. The only unique assertion here is the `GenerationId` check at line 275.

Consider merging the `GenerationId` assertion into Test 1:
```csharp
// At the end of Every_emitted_message_carries_the_runs_run_id:
var toolCall = messages.OfType<ToolsCallMessage>().Single();
toolCall.GenerationId.Should().Be(runGenerationId,
    "the BUG H1 GenerationId behavior must be preserved");
```

This eliminates ~30 lines of duplicated setup and a redundant stream execution. Not blocking — just a cleanup suggestion.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Incremental review of iterations 3-4 (streaming resume feature).

## Summary

The streaming resume implementation is **well-designed and correct**. The replay buffer in \MultiTurnAgentBase\ uses a clean locking pattern (\_replayLock\) that guarantees exactly-once delivery -- a subscriber joining mid-run gets buffered messages via replay XOR live channel, never both, never neither.

### What I checked:
- :white_check_mark: **Concurrency correctness** -- \_replayLock\ guards register+snapshot vs append+snapshot-targets atomically. No deadlock risk (no nested lock inversions).
- :white_check_mark: **Buffer lifecycle** -- RunAssignmentMessage resets; RunCompletedMessage deactivates and clears (memory freed between runs per self-review fix).
- :white_check_mark: **Client race guard** -- \	hreadId.value !== existingThreadId\ after awaiting \getRunState\ prevents cross-thread contamination on rapid switching.
- :white_check_mark: **Error recovery** -- Failed resume socket resets \isLoading\ (no stuck spinner). \getRunState\ failure silently returns.
- :white_check_mark: **Buffer cap** -- 10,000 messages with graceful degradation (log warning, persisted history covers completed).
- :white_check_mark: **Test coverage** -- 5 backend tests (replay, no-replay-after-complete, multi-subscriber, concurrent-publish exactly-once, buffer cap) + 6 client tests (reconnect, no-reconnect-idle, already-connected no-op, SSE skip, failed-open reset, stale-thread abort).
- :white_check_mark: **Build passes** (only pre-existing Scriban NU1902 warnings).

### Previous review comments (iterations 1-2):
Two threads from the initial review remain open (Finding 1: missing \GenerateReplyAsync\ leftover path test [MEDIUM], Finding 2: Test 4 duplication [LOW]). These target code unchanged in iterations 3-4. Consider addressing them before merge.

**Verdict: APPROVE** -- the streaming resume feature is solid.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** — Incremental Review (Iterations 5-6: Context Discovery Diagnostics)

## Build & Tests
✅ Build succeeds (only pre-existing Scriban NU1902 warnings)
✅ All 47 context-discovery tests pass

---

## New Finding

### [MEDIUM] Blocking async in `Harness.Dispose` (ContextDiscoveryEndToEndTests.cs:169-173)

`Harness` implements `IDisposable` and calls `.GetAwaiter().GetResult()` on `Pool.DisposeAsync()` and `Registry.DisposeAsync()`. This violates the project's coding standards (no blocking async calls). The test methods are already `async Task`, so `Harness` should implement `IAsyncDisposable` with `await using`.

### [LOW] `StubHandler` duplicated in 6+ test files

`RecordingMultiTurnAgent` was correctly extracted to `TestDoubles/` for DRY in this commit — `StubHandler` (identical in all files) deserves the same treatment.

### [LOW] `CreateRegistry` factory copy-pasted 4 times

Nearly identical factory methods across the new test files. A shared builder in `TestDoubles/` or a base fixture would cut the boilerplate.

---

## Design Notes (positive)

- The diagnostics architecture is clean: separation of ""what arrived"" (`ContextDiscoveryDiagnostics`) vs ""what's configured"" (registry `DiscoveryWebhookUrl`/`Enabled`) is well-thought-out.
- Thread-safety via `ConcurrentDictionary.AddOrUpdate` is correct.
- INFO log at `BuildDiscovery` correctly never leaks the shared secret.
- `GetActiveSessionIds()` defensively checks `_disposed` — good.
- The E2E + HTTP test approach (TestServer with real ASP.NET pipeline) provides strong regression coverage for the webhook path.

---

## Previous Comments (still active)
- **Thread 1 [MEDIUM]:** Missing test for `GenerateReplyAsync` leftover TextMessage path — unaddressed
- **Thread 2 [LOW]:** Test 4 near-duplicate of Test 1 — unaddressed

## Verdict: **APPROVE with suggestions**
No blocking issues. The MEDIUM (blocking async) is a test-only anti-pattern with no runtime risk in xUnit, but should be cleaned up to maintain project standards.

**Review by gautam-achieveai — COMMENTED** [review]
[Gautam's review bot] Incremental review (iterations 7-8): APPROVE - No new issues.

Changes Reviewed:
- useChat.ts (3-line fix): Stamps currentRunId onto runId-less live messages before merge-key computation, fixing the frozen-tool-pill duplicate bug on switch-away/back resume.
- useChatResume.test.ts: Two new regression tests (1-tool and 12-tool switch-back repros, RED without fix, GREEN with).
- MultiTurnAgentReplayTests.cs: New backend test pinning replay buffer carries tool_call + tool_call_result.
- CLAUDE.md: Documents the merge-key divergence gotcha with root cause, localizer, and prevention rules.

Assessment:
Fix is surgical and correct. Guard only stamps when a run is active AND the message lacks a runId. currentRunId is set from run_assignment which the replay buffer delivers first on resume. Providers that already send a runId are unaffected. Tests reproduce the exact user-reported bug at both 1-tool and 12-tool scale.

Prior Open Items (not blocking merge):
1. [MEDIUM] Missing test for GenerateReplyAsync leftover TextMessage path (iter 1-2)
2. [MEDIUM] Blocking async in test Harness.Dispose (iter 5-6)
3. [LOW] Test 4 near-duplicate of Test 1 (iter 1-2)
4. [LOW] StubHandler/CreateRegistry duplication across test files (iter 5-6)

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Incremental review of iteration 9 (d511dd3).

**Changes:** Playwright E2E test proving the tool-pill resume fix end-to-end, plus supporting `data-testid` selectors in `ConversationSidebar.vue` and a `ConversationItems()` UiHelper.

**Verdict: APPROVE** - no new issues.

The E2E test is well-constructed:
- Faithfully reproduces real Anthropic wire behavior (strips runId from tool-call WS frames via Playwright WebSocket routing)
- Guards against vacuous passes by asserting the backend run is genuinely in-flight at switch-back
- Covers both `test-anthropic` and `test` provider modes
- Follows existing E2E patterns (UiHelpers, PlaywrightFixture, data-testid contract)

**Prior unaddressed comments** (non-blocking, from earlier iterations):
- Thread 1 (MEDIUM): Missing test for `GenerateReplyAsync` leftover TextMessage path
- Thread 2 (LOW): Test 4 near-duplicate of Test 1

File: src/LmMultiTurn/MultiTurnAgentBase.cs [hunk 3]
```diff
@@ -793,11 +809,33 @@ public async IAsyncEnumerable<IMessage> SubscribeAsync(
             SingleWriter = false,
         });
 
-        _outputSubscribers[subscriberId] = channel;
-        Logger.LogDebug("Subscriber {SubscriberId} connected", subscriberId);
+        // Atomically register this subscriber AND snapshot the in-flight run's buffered messages,
+        // so a message published concurrently is delivered EITHER via this replay snapshot OR via
+        // the live channel below — never both, never neither. See `_replayLock` remarks.
+        IReadOnlyList<IMessage> replay;
+        lock (_replayLock)
```
Comment (gautam-achieveai) [inline src/LmMultiTurn/MultiTurnAgentBase.cs:828]: [Gautam's bot] [BLOCKER] **HIGH** (Performance / Streaming Reliability)

`SubscribeAsync` now registers a reconnecting subscriber in `_outputSubscribers` before it has replayed the buffered messages. While that client is still consuming replay, new live messages are queued into its bounded channel. If the reconnecting WebSocket/client is slow enough for that channel to fill, `PublishToSubscriberAsync` awaits `channel.Writer.WriteAsync(...)`, and `PublishToAllAsync` awaits that task along with every other subscriber.

That puts a replaying/slow reconnecting client on the live publisher hot path: one stalled resume can backpressure the active agent run and other subscribers. This is especially risky because the replay window can be up to 10,000 messages.

**Suggestion:** Decouple replay from live publication. For example, use a sequence/cursor-based replay where the subscriber catches up before being added to the live fan-out, replay a compact snapshot instead of all deltas, or drop/disconnect slow subscribers when their live channel is full rather than awaiting indefinitely in the publisher path. The key requirement is that a reconnecting replay consumer must not be able to block `PublishToAllAsync` for the run.

**Comment by gautam-achieveai** [comment]
[Gautam's bot] # PR Review: Bug fixes

## Summary
- Total files reviewed: 52 changed files (deep review against merge-base `6354177`)
- Findings: 0 Critical, 1 High, 1 Medium, 1 Low
- Context questions: 0 posted as inline questions
- Test coverage: strong overall; no blocking coverage gaps found in this pass
- Domain areas touched: .NET provider/multi-turn streaming, ASP.NET sample APIs/diagnostics, Vue composables, browser/unit tests

## Strengths
- The PR includes substantial regression coverage for the main behaviors: OpenAI Responses run-id stamping, multi-turn replay/resume, client resume behavior, and browser E2E proof for tool-pill resume.
- The replay implementation avoids `await` while holding `_replayLock`, which is the right direction for the concurrency-sensitive path.
- Diagnostics intentionally avoid retaining webhook bodies or shared secrets.

## High Issues
| # | File | Line | Blocker? | Issue | Fix |
|---|---|---:|---|---|---|
| 1 | `src/LmMultiTurn/MultiTurnAgentBase.cs` | 816 | Yes | A reconnecting subscriber is registered for live fan-out before it finishes replay. If it consumes replay slowly and its live channel fills, `PublishToAllAsync` waits on that subscriber and can stall the active stream and other subscribers. | Decouple replay consumers from live publication: sequence/cursor replay before live registration, compact snapshot replay, or drop/disconnect slow subscribers instead of awaiting indefinitely on the publisher path. |

## Medium Issues
| # | File | Line | Blocker? | Issue | Fix |
|---|---|---:|---|---|---|
| 2 | `src/LmMultiTurn/MultiTurnAgentBase.cs` | 43 | No | The replay cap is by message count only (`10_000`). Individual `IMessage`s can carry large text/tool/reasoning payloads, so active runs can retain substantial memory until completion despite the count cap. | Add an estimated byte budget/compaction strategy, truncate or skip very large payloads, or store compact replay state instead of every full message. |

## Low Issues
| # | File | Line | Blocker? | Issue | Fix |
|---|---|---:|---|---|---|
| 3 | PR metadata | - | No | Title “Bug fixes” undersells a broad multi-area diff and leaves reviewers inferring how all diagnostics/sandbox/tool-mode changes relate to #110. | Rename/update the description to explicitly connect each changed area to #110, or split/rebase unrelated carry-over work if any exists. |

## Testing Assessment
Coverage is generally strong: the PR adds/updates provider unit tests, replay contract tests, client resume tests, diagnostics/controller tests, and browser E2E coverage. I did not find a new blocking test gap in this pass.

## Security / EUII Review
No secret logging or obvious EUII leak was found in the inspected diagnostics path. The diagnostics endpoint reports session IDs and callback URL; that looks intended for local/sample diagnostics, but keep it gated appropriately if this sample is ever exposed beyond a trusted dev environment.

## Verdict
**REQUEST CHANGES**

The replay/resume feature needs a publisher-hot-path fix before merge. The memory-budget issue is non-blocking but should be addressed or explicitly accepted with rationale.

File: src/LmMultiTurn/MultiTurnAgentBase.cs [hunk 4]
```diff
@@ -26,6 +26,22 @@ public abstract class MultiTurnAgentBase : IMultiTurnAgent
...
+    // register-subscriber + buffer-snapshot (SubscribeAsync) atomically against the buffer-append +
+    // subscriber-snapshot (PublishToAllAsync), so a message published concurrently with a subscribe
+    // reaches that subscriber EXACTLY once (replay XOR live) — this holds even if publishes overlap
+    // (e.g. parallel tool-call results). Relative ordering of concurrently-published messages is not
+    // guaranteed (the channel writes happen outside the lock), exactly as before this change.
+    private readonly object _replayLock = new();
+    private readonly List<IMessage> _replayBuffer = [];
+    private bool _replayRunActive;
+    private bool _replayBufferTruncated;
+    private const int MaxReplayBufferSize = 10_000;
```
Comment (gautam-achieveai) [inline src/LmMultiTurn/MultiTurnAgentBase.cs]: [Gautam's bot] **MEDIUM** (Performance / Memory)

The replay buffer cap is by message count only (`MaxReplayBufferSize = 10_000`). Individual `IMessage` instances can carry large text deltas, tool arguments/results, reasoning content, or serialized payloads, so an active long-running run can still retain substantial memory until `RunCompletedMessage` clears the buffer. Multiple active conversations multiply that retained payload size.

**Suggestion:** Add an estimated byte budget in addition to the count cap, compact/truncate large payloads, or store replay state in a more compact form (for example, a sequence/cursor plus latest aggregate state) instead of retaining every full message object.

**Comment by gautam-achieveai** [comment]
[Gautam's bot] Review posting note: GitHub rejected the formal **Request changes** review action because this account is the PR author (`Review Can not request changes on your own pull request`). I posted the blocking/high issue and medium issue as inline comments and posted the summary as a regular PR comment with verdict **REQUEST CHANGES**.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Incremental review (iteration 10) - commit `d587731`

## Change: Sub-agent concurrency cap scaled to host

The `const 5` to `static readonly Math.Max(1, Math.Min(16, ProcessorCount - 2))` is a sensible change. The floor-at-1 prevents `SemaphoreSlim(0)` and the cap-at-16 bounds fan-out. Comment clearly explains rationale.

### Minor observations (non-blocking)

| # | Severity | Note |
|---|----------|------|
| 1 | LOW | Environment-dependent behavior (CI gets 1, dev gets 6-14) may affect test reproducibility if sub-agent timeout tests exist |
| 2 | LOW | No unit test for the ProcessorCount <= 2 floor-at-1 boundary; static readonly makes this hard to test - consider extracting to a method if the formula evolves |

### Prior open threads (unchanged files - carrying forward)
- **MEDIUM:** Missing test for GenerateReplyAsync leftover TextMessage path (iter 1)
- **MEDIUM:** Replay buffer memory cap is count-only (iter 7)

**Verdict: :white_check_mark: APPROVE** - no new blocking issues.

File: src/LmMultiTurn/MultiTurnAgentBase.cs (hunk 3 above)
Comment (gautam-achieveai) [reply src/LmMultiTurn/MultiTurnAgentBase.cs:828]: Fixed in 766d60a. The publisher no longer awaits subscriber writes: `PublishToAllAsync` now writes non-blocking (`TryWrite`) and **drops** any subscriber whose bounded channel is full (completing its channel so its `SubscribeAsync` enumerator ends and the client can reconnect → resume replays the in-flight run). A reconnecting/slow replay consumer can therefore never backpressure the live run or other subscribers. Regression test: `A_stalled_subscriber_neither_blocks_the_publisher_nor_starves_other_subscribers`.

File: src/LmMultiTurn/MultiTurnAgentBase.cs (hunk 4 above)
Comment (gautam-achieveai) [reply src/LmMultiTurn/MultiTurnAgentBase.cs]: Addressed in 766d60a. Added an estimated byte budget (`_maxReplayBufferBytes`, default 8 MiB via `EstimateMessageBytes`) alongside the count cap — whichever trips first stops buffering. Both caps are now constructor-configurable. Test: `Replay_buffer_is_capped_by_estimated_bytes_not_just_count`.

File: tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs (hunk 1 above)
Comment (gautam-achieveai) [reply tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs:155]: Added in 766d60a as `GenerateReplyAsync_leftover_text_carries_run_ids` (with a `TextDeltasOnlyStream` fixture: text deltas, no finalizing output_text.done). It drives the non-streaming `GenerateReplyAsync` leftover-TextMessage fix site and asserts RunId/ParentRunId/ThreadId — distinct from the streaming `StampRunIds` wrapper.

File: tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs (hunk 2 above)
Comment (gautam-achieveai) [reply tests/OpenAiResponsesProvider.Tests/OpenAiResponsesAgentRunIdTests.cs]: Done in 766d60a. Folded the `GenerationId` assertion into `Every_emitted_message_carries_the_runs_run_id` and removed the near-duplicate `Tool_call_carries_full_run_thread_and_parent_ids`.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Incremental review (iteration 11) - commit `766d60a`

## Summary

All 4 previously open threads have been **addressed** - great work on the non-blocking publisher redesign and byte-budget cap. The architecture (TryWrite + drop slow subscriber + reconnect-replays-from-buffer) is sound and well-documented.

However, the **LmMultiTurn project does not build** due to an analyzer error. This blocks merge.

## Findings

### [HIGH] Build failure - IDE0048 in `EstimateMessageBytes`

Lines 975 and 977 of `MultiTurnAgentBase.cs` trigger `IDE0048: Parentheses should be added for clarity` (treated as error in this project):

```csharp
// Current (fails build):
return baseOverhead + (t.Text?.Length ?? 0) * 2L;

// Fix:
return baseOverhead + ((t.Text?.Length ?? 0) * 2L);
```

Same fix needed on line 985 for the `ToolsCallMessage` loop body.

### [LOW] `EstimateMessageBytes` doesn't account for `ToolsCallResultMessage`

Tool results can carry large payloads (serialized function output), but currently fall through to `default: return baseOverhead`. Given the comment says "intentionally approximate", this is acceptable for now - just noting for awareness.

## Previous Threads
- Missing test for `GenerateReplyAsync` leftover path - **ADDRESSED** (new test added)
- Test 4 near-duplicate - **ADDRESSED** (folded into existing test, duplicate removed)
- BLOCKER: publisher backpressure - **ADDRESSED** (non-blocking TryWrite + subscriber drop)
- Unbounded replay memory - **ADDRESSED** (byte budget added, 8 MiB default)

## Verdict: REQUEST CHANGES (build failure blocks merge)

File: src/LmMultiTurn/MultiTurnAgentBase.cs [hunk 5]
```diff
@@ -820,39 +870,127 @@ public async IAsyncEnumerable<IMessage> SubscribeAsync(
...
+    /// per message (tool-call args are summed, which is bounded by the call count).
+    /// </summary>
+    private static long EstimateMessageBytes(IMessage message)
+    {
+        const long baseOverhead = 128;
+        switch (message)
         {
-            Logger.LogWarning(ex, "Error publishing to subscriber {SubscriberId}", subscriberId);
+            case TextUpdateMessage t:
+                return baseOverhead + (t.Text?.Length ?? 0) * 2L;
```
Comment (gautam-achieveai) [inline src/LmMultiTurn/MultiTurnAgentBase.cs]: **[Gautam's review bot]** [HIGH] **Build failure: IDE0048**$([char]10)$([char]10)This expression triggers `IDE0048: Parentheses should be added for clarity` (treated as error in this project's analyzer config).
$([char]10)`csharp
// Current (build error):
return baseOverhead + (t.Text?.Length ?? 0) * 2L;

// Fix - wrap the multiplication:
return baseOverhead + ((t.Text?.Length ?? 0) * 2L);
`$([char]10)$([char]10)Same issue on line 977 (`TextMessage` case) and line 985 (`ToolsCallMessage` loop). The `LmMultiTurn.Tests` project fails to build until this is resolved.

File: src/LmMultiTurn/MultiTurnAgentBase.cs (hunk 3 above)
Comment (gautam-achieveai) [reply src/LmMultiTurn/MultiTurnAgentBase.cs:828]: **[Gautam's review bot]** Verified: the non-blocking TryWrite + subscriber-drop design eliminates the backpressure vector. The new `A_stalled_subscriber_neither_blocks_the_publisher_nor_starves_other_subscribers` test exercises this exactly. Well done - this was the most critical fix in the PR.

File: src/LmMultiTurn/MultiTurnAgentBase.cs (hunk 4 above)
Comment (gautam-achieveai) [reply src/LmMultiTurn/MultiTurnAgentBase.cs]: **[Gautam's review bot]** Verified: byte budget (`_maxReplayBufferBytes`, default 8 MiB) is now in place alongside the count cap. Both are constructor-configurable and tested via `Replay_buffer_is_capped_by_estimated_bytes_not_just_count`. Minor note: `ToolsCallResultMessage` falls through to the 128-byte base estimate - could undercount for large tool results, but acceptable as documented ('intentionally approximate').

**Comment by gautam-achieveai** [comment]
[Gautam's review bot] **APPROVE** — Iteration 12 incremental review

### Changes reviewed (since `766d60a`)
1. **IDE0048 fix** (`MultiTurnAgentBase.cs` L975, L977) — outer parentheses added to resolve the build-blocking analyzer error. Build now passes cleanly (0 errors).
2. **Redundant `!` removal** (`EchoDirectivesEndToEndTests.cs` L89, L174) — null-forgiving operators after `Assert.NotNull()` were unnecessary; good cleanup.

### Previous feedback status
All blocking/medium findings from iterations 9-11 have been addressed:
- :white_check_mark: IDE0048 build failure (BLOCKING -> fixed this commit)
- :white_check_mark: Missing test for `GenerateReplyAsync` leftover path (added iter 11)
- :white_check_mark: Near-duplicate test removal (folded iter 11)
- :white_check_mark: Publisher backpressure / subscriber starvation (non-blocking TryWrite, iter 11)
- :white_check_mark: Unbounded replay buffer memory (byte budget, iter 11)

### Remaining note (non-blocking)
`EstimateMessageBytes` doesn't cover `ToolsCallResultMessage` payloads (they fall through to the 128-byte base overhead). Fine for v1 — worth a follow-up if large tool results become common.

**Verdict: LGTM. Ship it!** :rocket:

**Comment by gautam-achieveai** [comment]
[Gautam's review bot] **APPROVE** — Iteration 13 incremental review

Re-checked the latest delta at `8a80a690` with the requested parallel review pass. The only changes since `766d60a` are the IDE0048 clarifying parentheses in `MultiTurnAgentBase.cs:975,977` and IDE0370 null-forgiving cleanup in `EchoDirectivesEndToEndTests.cs:89,174`; test coverage, architecture, performance, schema compatibility, exception handling, duplication, simplification, feature-flag, EUII, temp-code, alignment, security, and client-focused agents found no new findings.

GitHub checks are now green (Build and test + browser-e2e), so the previous build/analyzer blocker is resolved. **Verdict: LGTM / no blocking findings.**

**Comment by gautam-achieveai** [comment]
[Gautam's review bot] Review posting note: GitHub rejected the formal **Approve** review action because this account is the PR author (`Review Can not approve your own pull request`). The re-review summary above records the effective verdict: **LGTM / no blocking findings**.
